// @fluencelabs/fluence/dist/internal/FluenceConnection.js
"use strict";
/*
 * Copyright 2020 Fluence Labs Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FluenceConnection = exports.PROTOCOL_NAME = void 0;
var libp2p_websockets_1 = __importDefault(require("libp2p-websockets"));
var libp2p_mplex_1 = __importDefault(require("libp2p-mplex"));
var libp2p_1 = __importDefault(require("libp2p"));
var it_length_prefixed_1 = require("it-length-prefixed");
var it_pipe_1 = __importDefault(require("it-pipe"));
var log = __importStar(require("loglevel"));
var Particle_1 = require("./Particle");
var libp2p_noise_1 = require("@chainsafe/libp2p-noise");
var filters_1 = require("libp2p-websockets/src/filters");
var Buffer_1 = __importDefault(require("./Buffer"));
exports.PROTOCOL_NAME = '/fluence/particle/2.0.0';
var FluenceConnection = /** @class */ (function () {
    function FluenceConnection() {
    }
    FluenceConnection.createConnection = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var res, transportKey, _a;
            var _b;
            var _this = this;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        res = new FluenceConnection();
                        transportKey = libp2p_websockets_1.default.prototype[Symbol.toStringTag];
                        _a = res;
                        return [4 /*yield*/, libp2p_1.default.create({
                                peerId: options.peerId,
                                modules: {
                                    transport: [libp2p_websockets_1.default],
                                    streamMuxer: [libp2p_mplex_1.default],
                                    connEncryption: [libp2p_noise_1.NOISE],
                                },
                                config: {
                                    transport: (_b = {},
                                        _b[transportKey] = {
                                            filter: filters_1.all,
                                        },
                                        _b),
                                },
                                dialer: {
                                    dialTimeout: options === null || options === void 0 ? void 0 : options.dialTimeoutMs,
                                },
                            })];
                    case 1:
                        _a._lib2p2Peer = _c.sent();
                        res._lib2p2Peer.handle([exports.PROTOCOL_NAME], function (_a) {
                            var connection = _a.connection, stream = _a.stream;
                            return __awaiter(_this, void 0, void 0, function () {
                                var _this = this;
                                return __generator(this, function (_b) {
                                    (0, it_pipe_1.default)(stream.source, (0, it_length_prefixed_1.decode)(), function (source) { var source_1, source_1_1; return __awaiter(_this, void 0, void 0, function () {
                                        var msg, particle, e_1_1, e_2;
                                        var e_1, _a;
                                        return __generator(this, function (_b) {
                                            switch (_b.label) {
                                                case 0:
                                                    _b.trys.push([0, 13, , 14]);
                                                    _b.label = 1;
                                                case 1:
                                                    _b.trys.push([1, 6, 7, 12]);
                                                    source_1 = __asyncValues(source);
                                                    _b.label = 2;
                                                case 2: return [4 /*yield*/, source_1.next()];
                                                case 3:
                                                    if (!(source_1_1 = _b.sent(), !source_1_1.done)) return [3 /*break*/, 5];
                                                    msg = source_1_1.value;
                                                    try {
                                                        particle = Particle_1.Particle.fromString(msg);
                                                        options.onIncomingParticle(particle);
                                                    }
                                                    catch (e) {
                                                        log.error('error on handling a new incoming message: ' + e);
                                                    }
                                                    _b.label = 4;
                                                case 4: return [3 /*break*/, 2];
                                                case 5: return [3 /*break*/, 12];
                                                case 6:
                                                    e_1_1 = _b.sent();
                                                    e_1 = { error: e_1_1 };
                                                    return [3 /*break*/, 12];
                                                case 7:
                                                    _b.trys.push([7, , 10, 11]);
                                                    if (!(source_1_1 && !source_1_1.done && (_a = source_1.return))) return [3 /*break*/, 9];
                                                    return [4 /*yield*/, _a.call(source_1)];
                                                case 8:
                                                    _b.sent();
                                                    _b.label = 9;
                                                case 9: return [3 /*break*/, 11];
                                                case 10:
                                                    if (e_1) throw e_1.error;
                                                    return [7 /*endfinally*/];
                                                case 11: return [7 /*endfinally*/];
                                                case 12: return [3 /*break*/, 14];
                                                case 13:
                                                    e_2 = _b.sent();
                                                    log.debug('connection closed: ' + e_2);
                                                    return [3 /*break*/, 14];
                                                case 14: return [2 /*return*/];
                                            }
                                        });
                                    }); });
                                    return [2 /*return*/];
                                });
                            });
                        });
                        res._relayAddress = options.relayAddress;
                        return [2 /*return*/, res];
                }
            });
        });
    };
    FluenceConnection.prototype.disconnect = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this._lib2p2Peer.stop()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    FluenceConnection.prototype.sendParticle = function (particle) {
        return __awaiter(this, void 0, void 0, function () {
            var conn, sink;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        particle.logTo('debug', 'sending particle:');
                        return [4 /*yield*/, this._lib2p2Peer.dialProtocol(this._relayAddress, exports.PROTOCOL_NAME)];
                    case 1:
                        conn = _a.sent();
                        sink = conn.stream.sink;
                        (0, it_pipe_1.default)(
                        // force new line
                        [Buffer_1.default.from(particle.toString(), 'utf8')], (0, it_length_prefixed_1.encode)(), sink);
                        return [2 /*return*/];
                }
            });
        });
    };
    FluenceConnection.prototype.connect = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a, e1_1, e, error;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this._lib2p2Peer.start()];
                    case 1:
                        _b.sent();
                        log.debug("dialing to the node with client's address: " + this._lib2p2Peer.peerId.toB58String());
                        _b.label = 2;
                    case 2:
                        _b.trys.push([2, 4, , 5]);
                        _a = this;
                        return [4 /*yield*/, this._lib2p2Peer.dial(this._relayAddress)];
                    case 3:
                        _a._connection = _b.sent();
                        return [3 /*break*/, 5];
                    case 4:
                        e1_1 = _b.sent();
                        e = e1_1;
                        if (e.name === 'AggregateError' && e._errors.length === 1) {
                            error = e._errors[0];
                            throw "Error dialing node " + this._relayAddress + ":\n" + error.code + "\n" + error.message;
                        }
                        else {
                            throw e;
                        }
                        return [3 /*break*/, 5];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    return FluenceConnection;
}());
exports.FluenceConnection = FluenceConnection;
