// @fluencelabs/fluence/dist/internal/Particle.js
"use strict";
/*
 * Copyright 2020 Fluence Labs Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Particle = void 0;
var uuid_1 = require("uuid");
var base64_js_1 = require("base64-js");
var loglevel_1 = __importDefault(require("loglevel"));
var utils_1 = require("./utils");
var Buffer_1 = __importDefault(require("./Buffer"));
var Particle = /** @class */ (function () {
    function Particle() {
        this.callResults = [];
    }
    Particle.createNew = function (script, ttlMs) {
        var res = new Particle();
        res.id = genUUID();
        res.script = script;
        res.ttl = ttlMs;
        res.data = Buffer_1.default.from([]);
        res.timestamp = Date.now();
        return res;
    };
    Particle.fromString = function (str) {
        var json = JSON.parse(str);
        var res = new Particle();
        res.id = json.id;
        res.initPeerId = json.init_peer_id;
        res.timestamp = json.timestamp;
        res.ttl = json.ttl;
        res.script = json.script;
        res.signature = json.signature;
        res.data = (0, base64_js_1.toByteArray)(json.data);
        return res;
    };
    Particle.prototype.getParticleContext = function () {
        return {
            particleId: this.id,
            initPeerId: this.initPeerId,
            timestamp: this.timestamp,
            ttl: this.ttl,
            signature: this.signature,
        };
    };
    Particle.prototype.actualTtl = function () {
        return this.timestamp + this.ttl - Date.now();
    };
    Particle.prototype.hasExpired = function () {
        return this.actualTtl() <= 0;
    };
    Particle.prototype.clone = function () {
        var res = new Particle();
        res.id = this.id;
        res.initPeerId = this.initPeerId;
        res.timestamp = this.timestamp;
        res.ttl = this.ttl;
        res.script = this.script;
        res.signature = this.signature;
        res.data = this.data;
        res.callResults = this.callResults;
        return res;
    };
    Particle.prototype.toString = function () {
        var particle = this;
        var payload = {
            action: 'Particle',
            id: particle.id,
            init_peer_id: particle.initPeerId,
            timestamp: particle.timestamp,
            ttl: particle.ttl,
            script: particle.script,
            // TODO: copy signature from a particle after signatures will be implemented on nodes
            signature: [],
            data: (0, base64_js_1.fromByteArray)(particle.data),
        };
        return JSON.stringify(payload);
    };
    Particle.prototype.logTo = function (level, message) {
        var fn;
        var data;
        switch (level) {
            case 'debug':
                fn = loglevel_1.default.debug;
                data = (0, utils_1.dataToString)(this.data);
                break;
            case 'error':
                fn = loglevel_1.default.error;
                break;
            case 'info':
                fn = loglevel_1.default.info;
                break;
            case 'trace':
                fn = loglevel_1.default.info;
                break;
            case 'warn':
                fn = loglevel_1.default.warn;
                break;
            default:
                return;
        }
        fn(message, (0, utils_1.jsonify)({
            id: this.id,
            init_peer_id: this.initPeerId,
            timestamp: this.timestamp,
            ttl: this.ttl,
            script: this.script,
            signature: this.signature,
            callResults: this.callResults,
            data: data,
        }));
    };
    return Particle;
}());
exports.Particle = Particle;
function genUUID() {
    return (0, uuid_1.v4)();
}
